filebeat/harvester/registry.go

```diff
@@ -39,9 +45,7 @@ func (r *Registry) Stop() {
 	close(r.done)
 
 	for _, hv := range r.harvesters {
-		r.wg.Add(1)
 		go func(h Harvester) {
-			r.wg.Done()
 			h.Stop()
 		}(hv)
 	}
```

Comment:
where is `r.wg.Wait()` executed? Having `r.wg.Wait()` after this for loop, we can wait for all harvester to finish. This will allow us to make h.Stop() non-blocking ->

e.g.
```
func (r *Registry) Stop() {
  r.Lock()                // <- the lock on Start and Stop guarantees stop can only be called after Start finishes
  defer r.Unlock() 

  close(r.done)

  for _, h := range r.harvesters {
    h.Stop()
  }

  r.wg.Wait()   // we wait here before releasing the lock to ensure start can not be executed before Stop finishes 
}
```


with 

```
 func (r *Registry) Start(h Harvester) {
  r.Lock()
  defer r.Unlock()

  // Make sure no new harvesters are started after stop was called
  if !r.active() {
    return
  }

  r.wg.Add(1)
  r.harvesters[h.ID()] = h
  go func() {
    // <- note: due to stop holding the mutex and waiting on `r.wg` the harvester is removed from 
    // the registry after r.wg.Done is called. This potentially races with the prospector not being able to re-start
    // the harvester. With the prospector trying to periodically start the harvester, the fact the race exists can be "ignored".
    defer r.remove(h) 

    defer r.wg.Done()
    h.Start()
  }
}

func (r *Registry) active() bool {
  select {
    case <-r.done:
      return false
    default:
      return true
  }  
}

func (h *Harvester) Stop() {
  // return immediately -> shutdown will be complete asynchronously.
  // Due to `r.wg` being decremented, only after the harvester `Start/Run` returned, the `registry.Stop` will block anyways.
  h.stopOnce.Do(func() {
 		close(h.done)
   })
}
```